opm/material/densead/Evaluation1.hpp

```diff
@@ -0,0 +1,416 @@
+// -*- mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+// vi: set et ts=4 sw=4 sts=4:
+/*
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+/*!
+ * \file
+ *
+ * \brief Representation of an evaluation of a function and its derivatives w.r.t. a set
+ *        of variables in the localized OPM automatic differentiation (AD) framework.
+ */
+
+#ifndef OPM_LOCAL_AD_EVALUATION_1_HPP
+#define OPM_LOCAL_AD_EVALUATION_1_HPP
+
+#include "Math.hpp"
+
+#include <opm/common/Valgrind.hpp>
+
+#include <dune/common/version.hh>
+
+#include <array>
+#include <cmath>
+#include <cassert>
+#include <iostream>
+#include <algorithm>
+
+namespace Opm {
+namespace DenseAd {
+/*!
+ * \brief Represents a function evaluation and its derivatives w.r.t. a fixed set of
+ *        variables.
+ */
+template <class ValueT>
+class Evaluation< ValueT, 1 >
```

Comment:
you do not need to specialize the whole class here, it is sufficient to specialize the member functions/operators that are required, a.k.a.,

```c++
template <class ValueT>
Evaluation<ValueT, 1>& Evaluation<ValueT, 1>::operator+=(const Evaluation<ValueT, 1>& b)
{
   data[0] += b.data[0];
   data[1] += b.data[1];
   return *this;
}
```

(I tested this on a little test program, so there may be reality may prevent this.) This would have the advantage that the specialized classes are automatically consistent with the unspecialised one.